blog/fb4/index.qmd

```diff
@@ -0,0 +1,162 @@
+---
+author: [Audrey Blarr, Jason Gyamfi, Gregory M. Kapfhammer]
+title: 'Mutation Analysis'
+date: '2023-10-10'
+date-format: long
+categories: [post, software engineering, fuzzing book]
+description: <em>How can we incorporate mutations into our code for further analysis?</em>
+toc: true
+page-layout: full
+---
+
+## Overview
+
+This is our fourth post about the [The Fuzzing
+Book](https://www.fuzzingbook.org/)! The goal of this post is to relate the
+material from the book to our progress on our
+[Chasten](https://github.com/AstuteSource/chasten) tool, with the hope that it
+will provide valuable lessons on how we approach designing and developing our
+own software. Part of this is guidance on automating tests, and how that may
+look in different coding languages &mdash; although the article's main focus is
+the Python programming language.
+
+## Summary
+
+Mutations, or artificial faults, can be injected into code to check whether or not a test suite will pick up on specific bugs. On subject programs there are two highlighted methods of running mutation analysis; MuFunctionAnalyzer (implementing mutation analysis for individual functions) and MuProgramAnalyzer (implementing mutation analysis for standalone programs containing test suites). 
+
+Here is an example of MuFunctionAnalyzer at work:
+
+```python
+for mutant in MuFunctionAnalyzer(gcd, log=True):
+>>>     with mutant:
+>>>         assert gcd(1, 0) == 1, "Minimal"
+>>>         assert gcd(0, 1) == 1, "Mirror"
+>>> mutant.pm.score()
+->  gcd_1
+<-  gcd_1
+Detected gcd_1 <class 'UnboundLocalError'> local variable 'c' referenced before assignment
+
+->  gcd_2
+<-  gcd_2
+Detected gcd_2 <class 'AssertionError'> Mirror
+```
+
+And MuProgramAnalyzer:
+
+```python
+class TestGCD(unittest.TestCase):
+>>>     def test_simple(self):
+>>>         assert cfg.gcd(1, 0) == 1
+>>> 
+>>>     def test_mirror(self):
+>>>         assert cfg.gcd(0, 1) == 1
+>>> for mutant in MuProgramAnalyzer('gcd', gcd_src):
+>>>     mutant[test_module].runTest('TestGCD')
+>>> mutant.pm.score()
+======================================================================
+FAIL: test_simple (__main__.TestGCD)
+----------------------------------------------------------------------
+Traceback (most recent call last):
+  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_78662/2565918356.py", line 3, in test_simple
+    assert cfg.gcd(1, 0) == 1
+AssertionError
+```
+
+Structural coverage, as discussed previously, may not be enough since an execution that produces a wrong output yet is unnoticed by the test suite is counted exactly the same as an execution that produces the right output. Ineffective tests may achieve 100% code coverage, yet inaccurately represent the severe lack of ability of the program to detect bugs:
```

Comment:
Where was this discussed previously? Since this is a web site, can you please provide a link?

Reply:
I provided a link to an earlier chapter of FuzzingBook called Code Coverage.